pkg/sql/alter_table.go

```diff
@@ -1740,6 +1493,412 @@ func (p *planner) updateFKBackReferenceName(
+			if err := params.p.tryRemoveFKBackReferences(
+				params.ctx, tableDesc, constraint, t.DropBehavior, nil,
+			); err != nil {
+				return nil, err
+			}
+		}
+	}
+	tableDesc.UniqueWithoutIndexConstraints = tableDesc.UniqueWithoutIndexConstraints[:sliceIdx]
+
+	// Drop check constraints which reference the column.
+	constraintsToDrop := make([]string, 0, len(tableDesc.Checks))
+	constraintInfo, err := tableDesc.GetConstraintInfo()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, check := range tableDesc.AllActiveAndInactiveChecks() {
+		if used, err := tableDesc.CheckConstraintUsesColumn(check, colToDrop.GetID()); err != nil {
+			return nil, err
+		} else if used {
+			if check.Validity == descpb.ConstraintValidity_Dropping {
+				// We don't need to drop this constraint, its already
+				// in the process.
+				continue
+			}
+			constraintsToDrop = append(constraintsToDrop, check.Name)
+		}
+	}
+
+	for _, constraintName := range constraintsToDrop {
+		err := tableDesc.DropConstraint(params.ctx, constraintName, constraintInfo[constraintName],
+			func(*tabledesc.Mutable, *descpb.ForeignKeyConstraint) error {
+				return nil
+			},
+			params.extendedEvalCtx.Settings,
+		)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if err := params.p.removeColumnComment(params.ctx, tableDesc.ID, colToDrop.GetID()); err != nil {
+		return nil, err
+	}
+
+	// Since we are able to drop indexes used by foreign keys on the origin side,
+	// the drop index codepaths aren't going to remove dependent FKs, so we
+	// need to do that here.
+	// We update the FK's slice in place here.
+	sliceIdx = 0
+	for i := range tableDesc.OutboundFKs {
+		tableDesc.OutboundFKs[sliceIdx] = tableDesc.OutboundFKs[i]
+		sliceIdx++
+		fk := &tableDesc.OutboundFKs[i]
+		if descpb.ColumnIDs(fk.OriginColumnIDs).Contains(colToDrop.GetID()) {
+			sliceIdx--
+			if err := params.p.removeFKBackReference(params.ctx, tableDesc, fk); err != nil {
+				return nil, err
+			}
+		}
+	}
+	tableDesc.OutboundFKs = tableDesc.OutboundFKs[:sliceIdx]
+
+	found := false
+	for i := range tableDesc.Columns {
+		if tableDesc.Columns[i].ID == colToDrop.GetID() {
+			tableDesc.AddColumnMutation(colToDrop.ColumnDesc(), descpb.DescriptorMutation_DROP)
+			// Use [:i:i] to prevent reuse of existing slice, or outstanding refs
+			// to ColumnDescriptors may unexpectedly change.
+			tableDesc.Columns = append(tableDesc.Columns[:i:i], tableDesc.Columns[i+1:]...)
+			found = true
+			break
+		}
+	}
+	if !found {
+		return nil, pgerror.Newf(pgcode.ObjectNotInPrerequisiteState,
+			"column %q in the middle of being added, try again later", t.Column)
+	}
+
+	return droppedViews, validateDescriptor(params.ctx, params.p, tableDesc)
+}
+
+func handleTTLStorageParamChange(
+	params runParams,
+	tn *tree.TableName,
+	tableDesc *tabledesc.Mutable,
+	before, after *catpb.RowLevelTTL,
+) error {
+	switch {
+	case before == nil && after == nil:
+		// Do not have to do anything here.
+	case before != nil && after != nil:
```

Comment:
I think there's some thorny cases around removing the column but keeping TTL. Also, can we get some testing on multiple clauses where you remove the column via a reset and then add it back with a set in the same statement and the same transaction?

Reply:
we will block allowing this behaviour